packages/core/src/tracing/instrumentation/messaging/rdkafka.js

```diff
@@ -0,0 +1,417 @@
+   * are covered, we make sure that the data is always an array, and loop through them at all times.
+   */
+  if (!Array.isArray(eventData)) {
+    eventData = [eventData];
+  }
+
+  if (isActive) {
+    eventData.forEach(messageData => {
+      const parentSpan = cls.getCurrentSpan();
+
+      if (!parentSpan) {
+        if (cls.tracingSuppressed()) {
+          const headers = addTraceLevelSuppression(messageData.headers);
+          messageData.headers = headers;
+          return originalEmit.apply(ctx, originalArgs);
+        } else {
+          let traceId;
+          let parentSpanId;
+          let level;
+
+          const instanaHeaders = (messageData.headers || []).filter(headerObject => {
+            const headerKey = Object.keys(headerObject)[0].toLowerCase();
+            return headerKey.indexOf('x_instana') > -1;
+          });
+
+          // flatten array to object
+          const instanaHeadersAsObject = {};
+
+          instanaHeaders.forEach(instanaHeader => {
+            const key = Object.keys(instanaHeader)[0];
+            instanaHeadersAsObject[key] = instanaHeader[key];
+          });
+
+          if (instanaHeaders.length) {
+            // Look for the the newer string header format first.
+            if (instanaHeadersAsObject[constants.kafkaTraceIdHeaderNameString]) {
+              traceId = String(instanaHeadersAsObject[constants.kafkaTraceIdHeaderNameString]);
+            }
+            if (instanaHeadersAsObject[constants.kafkaSpanIdHeaderNameString]) {
+              parentSpanId = String(instanaHeadersAsObject[constants.kafkaSpanIdHeaderNameString]);
+            }
+            if (instanaHeadersAsObject[constants.kafkaTraceLevelHeaderNameString]) {
+              level = String(instanaHeadersAsObject[constants.kafkaTraceLevelHeaderNameString]);
+            }
+
+            // Only fall back to legacy binary trace correlation headers if no new header is present.
+            if (traceId == null && parentSpanId == null && level == null) {
+              // The newer string header format has not been found, fall back to legacy binary headers.
+              if (instanaHeadersAsObject[constants.kafkaTraceContextHeaderNameBinary]) {
+                const traceContextBuffer = instanaHeadersAsObject[constants.kafkaTraceContextHeaderNameBinary];
+                if (Buffer.isBuffer(traceContextBuffer) && traceContextBuffer.length === 24) {
+                  const traceContext = tracingUtil.readTraceContextFromBuffer(traceContextBuffer);
+                  traceId = traceContext.t;
+                  parentSpanId = traceContext.s;
+                }
+              }
+              level = readTraceLevelBinary(instanaHeadersAsObject);
+            }
+            removeInstanaHeadersFromMessage(messageData);
+          }
+
+          if (event === 'data' || event === 'error') {
+            cls.ns.runAndReturn(function () {
+              if (level !== '1') {
```

Comment:
We should also allow for the level header not to be present at all and assume level 1 as default. This check should probably more be like `level === 0` instead of `level !== 1`.

Reply:
I'll fix that cc @willianpc

Reply:
Good point!